### PR TITLE
Feature/#95 improve enemy steering behaviour

### PR DESCRIPTION
**Pull request by mserier**

### CODE REVIEW

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 1]
```diff
@@ -6,95 +6,54 @@
...
     private bool _lost = true;
+    private readonly float _rotationSpeed = 2f;
+    private readonly float _potentialEnergyRange = 0.8f;
 
-    void Start()
+    private void Awake()
     {
-        _energyCharge = EnergyCapacity;
-        _rigidbody = GetComponent<Rigidbody>();
+        PathFollower = new GameObject("PathFollower").AddComponent(typeof(PathFollower)) as PathFollower;
```
Comment (hassanAlAwwadi) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]:
```suggestion
        PathFollower = new GameObject("PathFollower").AddComponent<PathFollower>();
```

File: src/Assets/LightSkyDefense/Resources/Shaders/TransparentCullFix.shader [hunk 2]
```diff
@@ -1,52 +1,52 @@
-﻿Shader "Custom/TransparentCullFix"
+Shader "Custom/TransparentCullFix"
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Resources/Shaders/TransparentCullFix.shader]: This file can be removed

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 3]
```diff
@@ -241,33 +199,59 @@ private void OnTriggerEnter(Collider collider)
...
+    private void EnergyBehaviour()
+    {
+        // Calculate energy potential
+        _potentialEnergy = _potentialEnergyRange - Vector3.Distance(transform.position, _pathPoints[PathFollower.PathPointIndex].Position);
 
-        if (_waypointIndex == (pathPoints.Length - _lookAheadDistance - 1))
+        if (_potentialEnergy >= 0)
         {
-            Finish();
+            Charge(0.1f);
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]: The charge and discharge rate should be configurable

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 4]
```diff
@@ -241,33 +199,59 @@ private void OnTriggerEnter(Collider collider)
...
+
+            if (_potentialEnergy < _potentialEnergyRange / 2)
+            {
+                _lost = true;
+            }
+        }
+
+        if (_energyCharge < 0)
+        {
+            Damage(30);
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]: This damage should be configurable

File: src/Assets/LightSkyDefense/Scripts/Path.cs [hunk 5]
```diff
@@ -8,8 +8,8 @@ public class Path : MonoBehaviour
 {
     public class PathPoint
     {
-        public Vector3 Position;
-        public Vector3 DirectionVector;
+        public Vector3 Position;// Tail
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Path.cs]:
```suggestion
        [Tooltip("Tail")]
        public Vector3 Position;
```

File: src/Assets/LightSkyDefense/Scripts/Path.cs [hunk 6]
```diff
@@ -8,8 +8,8 @@ public class Path : MonoBehaviour
...
+        public Vector3 DirectionVector;// Translation Vector
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Path.cs]:
```suggestion
        [Tooltip("Translation Vector")]
        public Vector3 DirectionVector;
```

File: src/Assets/LightSkyDefense/Scripts/SteeringBehaviours/Aggregates/EnemySteeringBehaviour.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        _enemy = GetComponent<Enemy>();
+    }
+
+    void FixedUpdate()
+    {
+        var targetPosition = _enemy.PathFollower.transform.position + _enemy.PathFollower.offsetTranslation;
+
+        var steeringForce = Arrive.Calculate(transform.position, _enemy.Rigidbody.velocity, targetPosition, MaxSpeed);
+
+        _enemy.Rigidbody.AddForce(Vector3.ClampMagnitude(steeringForce, MaxSpeed));
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/SteeringBehaviours/Aggregates/EnemySteeringBehaviour.cs]: Arrive already clamps to max speed

```suggestion
        _enemy.Rigidbody.AddForce(steeringForce,);
```

File: src/Assets/LightSkyDefense/Resources/Shaders/TransparentCullFix.shader (hunk 2 above)
Comment (mserier) [reply src/Assets/LightSkyDefense/Resources/Shaders/TransparentCullFix.shader]: removed

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs (hunk 4 above)
Comment (mserier) [reply src/Assets/LightSkyDefense/Scripts/Enemy.cs]: this was a soft death taking a maximum of 4 ticks, replaced with Explode()

File: src/Assets/LightSkyDefense/Scripts/Path.cs (hunk 5 above)
Comment (mserier) [reply src/Assets/LightSkyDefense/Scripts/Path.cs]: Not a monobehaviour

File: src/Assets/LightSkyDefense/Scripts/Path.cs (hunk 6 above)
Comment (mserier) [reply src/Assets/LightSkyDefense/Scripts/Path.cs]: Not a monobehaviour

File: src/Assets/LightSkyDefense/Scripts/SteeringBehaviours/Aggregates/EnemySteeringBehaviour.cs (hunk 7 above)
Comment (mserier) [reply src/Assets/LightSkyDefense/Scripts/SteeringBehaviours/Aggregates/EnemySteeringBehaviour.cs]: fixed

**Pull request merged**